src/enclave-agent/src/main.rs

```diff
@@ -0,0 +1,181 @@
+use std::sync::Arc;
+
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use clap::{crate_authors, crate_version, App, Arg};
+use image_rs::config::ImageConfig;
+use image_rs::image::ImageClient;
+use image_rs::snapshots;
+use kata_sys_util::validate;
+use protocols::{image, image_ttrpc};
+use tokio::signal::unix::{signal, SignalKind};
+use tokio::sync::Mutex;
+use ttrpc::asynchronous::Server;
+use ttrpc::{self, error::get_rpc_status as ttrpc_error};
+
+const CONTAINER_BASE: &str = "/run/enclave-cc/containers";
+const CC_IMAGE_WORK_DIR: &str = "/run/image/";
+
+// TODO: will replace with unix socket
+const SOCK_ADDR: &str = "tcp://0.0.0.0:7788";
+
+struct ImageService {
+    image_client: Arc<Mutex<ImageClient>>,
+}
+
+impl ImageService {
+    fn new() -> Self {
+        env::set_var("CC_IMAGE_WORK_DIR", CC_IMAGE_WORK_DIR);
+        let new_config = ImageConfig {
+            default_snapshot: snapshots::SnapshotType::OcclumUnionfs,
+            ..Default::default()
+        };
+        Self {
+            image_client: Arc::new(Mutex::new(ImageClient {
+                config: new_config,
+                ..Default::default()
+            })),
+        }
+    }
+
+    async fn pull_image(&self, req: &image::PullImageRequest) -> Result<String> {
+        let image = req.get_image();
+        let mut cid = req.get_container_id().to_string();
+
+        if cid.is_empty() {
+            let v: Vec<&str> = image.rsplit('/').collect();
+            if !v[0].is_empty() {
+                // ':' have special meaning for umoci during upack
+                cid = v[0].replace(":", "_");
+            } else {
+                return Err(anyhow!("Invalid image name: {:?}", image));
+            }
+        } else {
+            validate::verify_id(&cid)?;
+        }
+
+        let keyprovider_config = Path::new("/etc").join("ocicrypt_keyprovider_native.conf");
+        if !keyprovider_config.exists() {
+            let config = r#"
+            {
+                "key-providers": {
+                    "attestation-agent": {
+                        "native": "attestation-agent"
+                    }
+                }
+            }
+            "#;
+            File::create(&keyprovider_config)?.write_all(config.as_bytes())?;
+        }
+        std::env::set_var("OCICRYPT_KEYPROVIDER_CONFIG", keyprovider_config);
+
+        let mut config: String = String::new();
+        let args: Vec<String> = env::args().collect();
+        // If config file specified in the args, read contents from config file
+        let config_position = args.iter().position(|a| a == "--config" || a == "-c");
+        if let Some(config_position) = config_position {
+            if let Some(config_file) = args.get(config_position + 1) {
+                config = fs::read_to_string(config_file).expect("Config file not found");
+            } else {
+                panic!("The config argument wasn't formed properly: {:?}", args);
+            }
+        }
```

Comment:
Please can you use the `clap` routines (already added on this PR) for argument parsing.

Reply:
> Please can you use the `clap` routines (already added on this PR) for argument parsing.

Sorry, I can not get your point, what is `clap routines` ?
I think it is a better way to parse argument when used.

Reply:
Well, I'm honestly rather confused. This PR:

- Adds a `src/enclave-agent/src/main.rs`, which defines a `main()` function.
- `main()` parses CLI arguments using `clap::App::new()`, and then starts the server.
- The server creates an instance of `ImageService` which defines a method called `pull_image()`.
- (_And now the bit I don't understand_): `pull_image()` *also* parses CLI options "manually":
  ```rust
  let args: Vec<String> = env::args().collect(); // get CLI args

  // Parse them
  let config_position = args.iter().position(|a| a == "--decrypt-config" || a == "-c");

  // ...
  ```

Hence, my questions are:

1) Why does `pull_image()` also need to parse CLI arguments? Why can't this all be handled in `main()` and passed to the `ImageService` service as a config struct or similar?
2) If you really, really need to do global CLI parsing in the server code, you can do it with clap. That would be consistent with `main()` and avoid "custom parsing" (why bother? we've got clap to help us! ;)
   ```rust
   async fn pull_image(&self, req: &image::PullImageRequest) -> Result<String> {
       let args: Vec<String> = env::args().collect();

       let matches = App::new("ocicrypt-config-parser").blah.blah.blah.get_matches_from(args);

       // ...
   }
   ```

Reply:
@jodh-intel  Hi Jiames, The config is not a global configuration, just in the pull image filed, so it's better to put the config file into the function filed rather than be global